### PR TITLE
Adding docs on using tokio task_local to share state

### PR DESCRIPTION
**Pull request by Gaelik-git**
I stumbled upon this [discussion](https://github.com/tokio-rs/axum/discussions/2631) for using tokio task_local to share state between handlers and middleware.

Is it okay to add some docs, would you prefer to have some test or example to show the behaviour ?

### CODE REVIEW

File: axum/src/docs/middleware.md [hunk 1]
```diff
@@ -535,6 +535,60 @@ let app = Router::new()
 automatically moved to response extensions. You need to manually do that for the
 extensions you need.
 
+State can also be passed from middleware to handlers using [tokio task_local]:
```
Comment (jplatte) [inline axum/src/docs/middleware.md:538]:
```suggestion
State can also be passed from middleware to handlers using
[tokio's `task_local` macro](https://docs.rs/tokio/1/tokio/macro.task_local.html):
```

File: axum/src/docs/middleware.md [hunk 2]
```diff
@@ -535,6 +535,60 @@ let app = Router::new()
...
+    http::StatusCode,
+    middleware::{self, Next},
+    response::{IntoResponse, Response},
+    routing::get,
+    Router,
+};
+use tokio::task_local;
+
+#[derive(Clone)]
+struct CurrentUser {/* ... */}
```
Comment (jplatte) [inline axum/src/docs/middleware.md:552]:
```suggestion
struct CurrentUser {
    // ...
}
```

File: axum/src/docs/middleware.md [hunk 3]
```diff
@@ -535,6 +535,60 @@ let app = Router::new()
...
+    let auth_header = req
+        .headers()
+        .get(http::header::AUTHORIZATION)
+        .and_then(|header| header.to_str().ok());
+
+    let auth_header = if let Some(auth_header) = auth_header {
+        auth_header
+    } else {
+        return Err(StatusCode::UNAUTHORIZED);
+    };
```
Comment (jplatte) [inline axum/src/docs/middleware.md:568]:
```suggestion
    let auth_header = req
        .headers()
        .get(http::header::AUTHORIZATION)
        .and_then(|header| header.to_str().ok())
        .ok_or(StatusCode::UNAUTHORIZED)?;
```

**Review by jplatte — COMMENTED** [review]
Thanks for the PR! I think it makes sense to showcase this in the docs, but in addition to the small improvement suggestions below, there are two more general things:

- I think this documentation belongs in the main documentation, in the "Sharing state with handlers" section
- To me, the real important aspect about this is that unlike other state sharing mechanisms, this one actually also works to share state with `IntoResponse` impls; I think the docs should call this out

**Comment by jplatte** [comment]
Hey @Gaelik-git, are you still interested in getting back to this?

**Comment by Gaelik-git** [comment]
Yep sorry for the late response, I moved and modified the example for it to include the IntoReponse implementation
But my example does not make much sense I think, devs will still get the point I think but if you have other sugestion do not hesitate

File: axum/src/lib.rs [hunk 4]
```diff
@@ -293,6 +293,68 @@
 //! The downside to this approach is that it's a little more verbose than using
 //! [`State`] or extensions.
 //!
+//! ## Using [tokio's `task_local` macro](https://docs.rs/tokio/1/tokio/macro.task_local.html):
+//!
+//! This allows to share state with IntoResponse implementations.
```
Comment (jplatte) [inline axum/src/lib.rs:298]:
```suggestion
//! This allows to share state with `IntoResponse` implementations.
```

File: axum/src/lib.rs [hunk 5]
```diff
@@ -293,6 +293,68 @@
...
+//! }
+//!
+//! async fn handler() -> UserResponse {
+//!     UserResponse
+//! }
+//!
+//! let app: Router = Router::new()
+//!     .route("/", get(handler))
+//!     .route_layer(middleware::from_fn(auth));
+//!
```
Comment (jplatte) [inline axum/src/lib.rs:355]: The codeblock will look better without an empty line at the end.

```suggestion
```

**Review by jplatte — APPROVED** [review]
Thanks!

**Comment by Gaelik-git** [comment]
glad I could contribute

**Pull request merged**